### PR TITLE
Adding ETCI2021 datamodule and trainer

### PR DESCRIPTION
**Pull request by calebrob6**
Relevant links:
- https://competitions.codalab.org/competitions/30440#results
- https://arxiv.org/abs/2107.08369
- https://github.com/sidgan/ETCI-2021-Competition-on-Flood-Detection

### CODE REVIEW

File: torchgeo/trainers/segmentation.py [hunk 1]
```diff
@@ -77,19 +77,28 @@ def __init__(self, **kwargs: Any) -> None:
             in_channels: Number of channels in input image
             num_classes: Number of semantic classes to predict
             loss: Name of the loss function
+            ignore_zeros: Whether to ignore the "0" class value in the loss and metrics
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py:80]: Why not call it "ignore_index" and make it an int that defaults to zero? That gives the most flexibility.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Needs to be added to docs

File: torchgeo/datasets/etci2021.py [hunk 2]
```diff
@@ -254,3 +257,169 @@ def _download(self) -> None:
             filename=self.metadata[self.split]["filename"],
             md5=self.metadata[self.split]["md5"] if self.checksum else None,
         )
+
+    @staticmethod
+    def plot(sample: Dict[str, Tensor]) -> None:
```
[the remark this answers is not in the document]
Comment (calebrob6) [reply torchgeo/datasets/etci2021.py:262]: lets factor out this conversation to an issue

File: torchgeo/datasets/etci2021.py [hunk 3]
```diff
@@ -254,3 +257,169 @@ def _download(self) -> None:
...
+        """Plot a sample from the dataset.
+
+        Args:
+            sample: a sample returned by __getitem__
```
[the remark this answers is not in the document]
Comment (calebrob6) [reply torchgeo/datasets/etci2021.py]: yes, just the first

File: torchgeo/trainers/__init__.py [hunk 4]
```diff
@@ -4,16 +4,28 @@
...
+from .landcoverai import LandCoverAISegmentationTask
+from .naipchesapeake import NAIPChesapeakeSegmentationTask
 from .regression import RegressionTask
+from .resisc45 import RESISC45ClassificationTask
 from .segmentation import SemanticSegmentationTask
+from .so2sat import So2SatClassificationTask
 
 __all__ = (
     "BYOLTask",
+    "ChesapeakeCVPRSegmentationTask",
```
[the remark this answers is not in the document]
Comment (calebrob6) [reply torchgeo/trainers/__init__.py:19]: I disagree, you wrote deprecated on the classes, however there is not a clear path forward for how to get rid of the functionality. Until there is, I don't think it make sense to hide this functionality.

File: torchgeo/datasets/etci2021.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/datasets/etci2021.py:262]: #116 would be a good start

File: torchgeo/trainers/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/trainers/__init__.py:19]: I don't see GPU data augmentation as a requirement, but as something that would be nice to have. I started a conversation on this in https://github.com/PyTorchLightning/pytorch-lightning/issues/10469. Until then, I think we should stick to CPU data augmentation. Augmentation is much faster on the GPU, but I haven't yet seen whether data augmentation is actually the bottleneck or whether most of the time is spent in model training.

[the remark this answers is not in the document]
Comment (calebrob6) [reply torchgeo/trainers/__init__.py:19]: If you __reallly__ want to get rid of them, you could factor out the per dataset augmentations into their own modules (these could live inside the datamodules of you want), have all the generic tasks take a nn.Module argument that does augmentation given a sample, use an identity augmentation for all the datasets that don't have augmentations, and keep track of what goes where with config files.

[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/trainers/__init__.py:19]: Looks like we can easily put them in a DataModule: https://github.com/PyTorchLightning/pytorch-lightning/issues/10469

[the remark this answers is not in the document]
Comment (calebrob6) [reply torchgeo/trainers/__init__.py:19]: Great!! I will revert this part as there is a clear path moving forward

File: torchgeo/trainers/segmentation.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/trainers/segmentation.py:80]: Because nn.CrossEntropyLoss uses -1000 as a None value so there is not an easy way to not ignore.

**Comment by calebrob6** [comment]
@adamjstewart how do you want to test the plot methods?

File: torchgeo/trainers/segmentation.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/trainers/segmentation.py:80]: I'm not sure I understand. I'm saying to rename the `ignore_zeros` parameter to `ignore_index` and have it accept an int, not to remove the parameter entirely.

File: docs/conf.py [hunk 5]
```diff
@@ -59,6 +59,7 @@
     ("py:class", ".."),
     # TODO: can't figure out why this isn't found
     ("py:class", "LightningDataModule"),
+    ("py:class", "torchvision.datasets.folder.ImageFolder"),
```
Comment (adamjstewart) [inline docs/conf.py:62]: This can be removed now

File: tests/datasets/test_etci2021.py [hunk 6]
```diff
@@ -72,7 +72,10 @@ def test_getitem(self, dataset: ETCI2021) -> None:
             assert x["mask"].shape[0] == 1
 
     def test_len(self, dataset: ETCI2021) -> None:
-        assert len(dataset) == 2
+        if dataset.split == "train":
+            assert len(dataset) == 12
+        else:
+            assert len(dataset) == 4
```
Comment (adamjstewart) [inline tests/datasets/test_etci2021.py:78]: Would it be possible to reduce the number of test files (dataset length)? There are quite a lot.

File: torchgeo/trainers/etci2021.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+import kornia.augmentation as K
+import torch
+from torch import Tensor
+
+from .segmentation import SemanticSegmentationTask
+
+
+# TODO: move this functionality into SemanticSegmentationTask and remove this class
+class ETCI2021SemanticSegmentationTask(SemanticSegmentationTask):
```
Comment (adamjstewart) [inline torchgeo/trainers/etci2021.py:16]: We can move these augmentations to the DataModule and get rid of this class now

**Comment by adamjstewart** [comment]
> @adamjstewart how do you want to test the plot methods?

I would personally be fine with a simple test that just runs `plot` and makes sure it doesn't crash, but if you want to write a more detailed test you can check the figure object for all of the expected attributes. I once wrote a bunch of matplotlib auto-grading stuff for a class I was TAing.

File: torchgeo/trainers/segmentation.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/trainers/segmentation.py:80]: I know, but say I don't want to ignore anything -- I can't pass None because nn.CrossEntropyLoss behaves differently than everything else that accepts ignore_index

File: torchgeo/trainers/etci2021.py (hunk 7 above)
Comment (calebrob6) [reply torchgeo/trainers/etci2021.py:16]: out of scope of this PR for sure

**Comment by calebrob6** [comment]
Relevant bit for later
```
        x = batch["image"]
        y = batch["mask"]
        with torch.no_grad():
            # Kornia expects masks to be [B, C, H, W] format with float values while
            # torch losses like cross entropy expect masks to be in [B, H, W] format
            # with long values. We'll just assume that samples are in a format that
            # torch expects and adjust accordingly if we need to do augmentations.
            x, y = self.train_augmentations(x, y.unsqueeze(1).float())
        y = y.long().squeeze()
```

**Comment by calebrob6** [comment]
I removed the trainer and reset the test data as I can't be bothered to jump through all these hoops.

Note, this dataset **does** need augmentations as training will overfit to the train split (the test set is sampled from a different geographic location and those inputs are likely out-of-distribution to the training data).

The above comment is the relevant part to add to `train_step` for augmentation that explains what kornia augmentations are expecting in terms of dimensions.

**Review by adamjstewart — APPROVED** [review]
